sno/pk_generation.py

```diff
@@ -144,25 +184,138 @@ def _init_schema(self):
+        # once we know the full list of inserts and deletes, then we can reassign PKs from the deletes, so that they
+        # become edits.
+        buffered_inserts = []
+        buffered_insert_limit = self.similarity_detection_insert_limit
+
+        for orig_feature in self.delegate.features():
+            feature = {self.primary_key: None, **orig_feature}
+            feature_hash = self.schema.hash_feature(feature, without_pk=True)
+
+            pks = hash_to_unassigned_pks.get(feature_hash)
+            reassigned_pk = pks.pop(0) if pks else None
+
+            if reassigned_pk is not None:
+                # This feature is exactly the same as a historical one that had a PK,
+                # and that PK has not yet been assigned this import. We re-assign it now.
+                feature[self.primary_key] = reassigned_pk
+                yield feature
+
+            elif buffered_insert_limit > 0:
+                # New feature, but we don't assign it a PK just yet.
+                # We buffer this feature for now - maybe we'll find a similar one from among the deleted
+                # features later, which we can reuse the PK for, making this an edit.
+                # We can do this once we have the full list of new and deleted features).
+                buffered_inserts.append(feature)
+
+                if len(buffered_inserts) > buffered_insert_limit:
+                    # Too many inserts - give up on finding similar ones from among the deletes.
+                    yield from self._assign_pk_range(buffered_inserts, next_new_pk)
+                    next_new_pk += len(buffered_inserts)
+                    buffered_inserts = []
+                    buffered_insert_limit = 0
+
+            else:
+                # New feature. Assign it a new PK and yield it.
+                yield self._assign_pk(feature, next_new_pk, feature_hash=feature_hash)
+                next_new_pk += 1
+
+        if buffered_inserts:
+
+            # Look for matching inserts-deletes - reassign the PK from the delete, treat is as an edit:
+            yield from self._match_similar_features(
+                self._find_deleted_features(hash_to_unassigned_pks), buffered_inserts
+            )
+            # Just assign new PKs to those we couldn't find a match for.
+            yield from self._assign_pk_range(buffered_inserts, next_new_pk)
+
+    def _assign_pk_range(self, features, pk):
+        for feature in features:
+            yield self._assign_pk(feature, pk)
+            pk += 1
+
+    def _assign_pk(self, feature, pk, feature_hash=None):
+        if feature_hash is None:
+            feature_hash = self.schema.hash_feature(feature, without_pk=True)
+
+        feature[self.primary_key] = pk
         self.pk_to_hash[pk] = feature_hash
-        return pk
+        return feature
+
+    def _match_similar_features(self, old_features, new_features):
```

Comment:
I'd rename this to `_match_and_remove_similar_features` to make clear that it mutates both lists

Reply:
Chose _match_similar_features_and_remove